Set up Acceptance Testing Suite

karl-thomas opened the pull request.

karl-thomas committed:
add http_server_spec module

karl-thomas committed:
make it so that main will use the first cli argument as the port

karl-thomas committed:
refactor to use global for crlfs

karl-thomas committed:
seperate echo code and http code

karl-thomas committed:
get echo server integration tests runnignwith new folder structure

karl-thomas committed:
make a switch between http and echo

karl-thomas committed:
make acceptance test for simple_get pass

karl-thomas committed:
Rename HTTPserver.java to HTTPServer.java

lpenzey commented on src/main/java/com/karl/http/HTTPHandler.java:
```diff
@@ -0,0 +1,32 @@
+      final BufferedReader requestReader = socket.getReader();
+      String requestString = "";
+      while (requestReader.ready()) {
+        requestString = requestString + requestReader.readLine();
```
Replacing the `+` concatenation with `StringBuilder.append(requestReader.readLine())` is going to be a lot more performant.

karl-thomas replied:
No doubt 👍

lpenzey commented on src/main/java/com/karl/http/HTTPServer.java:
```diff
@@ -0,0 +1,33 @@
+import com.karl.wrappers.WrappedServerSocket;
+import com.karl.wrappers.WrappedSocket;
+
+public class HTTPServer {
```
This is almost identical to `EchoServer` aside from the `handler`- is there a way to get one server class to handle multiple `handlers`?

karl-thomas replied:
Absolutely, this is something i want to do, this is just the first thing i settled on.

lpenzey commented:
`README` 😬

lpenzey commented on src/test/java/com/karl/http/HTTPHandlerTest.java:
```diff
@@ -0,0 +1,23 @@
+  @Test
+  @DisplayName("server will respond with a 200 okay when a Get request is made to /simple_get")
+  public void respond200toSimpleGet() throws IOException {
+    String simpleGet = "GET /simple_get HTTP/1.1" + Globals.CRLF + Globals.CRLF;
```
what happens when a request comes in that doesn't adhere to the `HTTP` spec?

ranizilpelwar commented on src/main/java/com/karl/Main.java:
```diff
@@ -2,12 +2,33 @@
+    }
+
+    WrappedServerSocket socket = new ServerSocketWrapper(port);
+    if (serverType.equals("echo")) {
```
What happens if serverType is not one of the two? 
Possible refactor to a switch statement for readability.

karl-thomas committed:
update README.md

karl-thomas committed:
get http acceptance tests to run on mvn verify

karl-thomas committed:
set up http acceptance testing

run acceptance suite on integration-test phase

use checkout action to get submodules

add access token to actions

switch to github token for actions

change token to be a personal access token

remove submodule action from push workflow

remove part of the working dir for acceptance test execution

update execution id's

modify acceptance test exectution

try to back up a directory in ci

check pwd before testing

run ls before testing

list contents of submodule dir in action

remove back up from wirking dir in acceptance test execution

update working dir

set up ruby in action

set up dependencies on CI

fix typo

set up caching for ruby gems

set up http execution

test the specific scenario i am working on

karl-thomas committed:
refactor to use string builder, use hamcrest mathcer

karl-thomas committed:
update route to path

karl-thomas committed:
start work on request builder class

karl-thomas committed:
remove request string parsing to request builder

karl-thomas committed:
rename getter methods

karl-thomas committed:
create methodIs function to match against the request method

karl-thomas committed:
follow convention a bit better with the request builder class

karl-thomas committed:
add ServerType enum to enforce the right server type

karl-thomas committed:
create handler factories in order to generalize server code

karl-thomas committed:
update names of classes to bring in a socket interface and mock

karl-thomas committed:
add a read until function to help read requests better

karl-thomas committed:
add conveniece methods to request builder
this is in order to interaction with a client connection directly, after having properly mocked it out

karl-thomas committed:
fix breaking test

lpenzey commented on src/main/java/com/karl/http/HTTPHandlerFactory.java:
```diff
@@ -0,0 +1,10 @@
+import com.karl.ProtocolFactoryable;
+import com.karl.wrappers.Connectable;
+
+public class HTTPHandlerFactory implements ProtocolFactoryable {
```
nice 🎉

karl-thomas replied:
nice 👍

lpenzey commented on src/main/java/com/karl/http/HTTPMethod.java:
```diff
@@ -0,0 +1,14 @@
+package com.karl.http;
+
+public enum HTTPMethod {
+  OPTIONS, GET, HEAD, POST, PUT, DELETE, INVALID;
```
Nice job locking down eligible methods 👍

ranizilpelwar replied:
+1

lpenzey commented on src/main/java/com/karl/http/HTTPRequestBuilder.java:
```diff
@@ -0,0 +1,49 @@
+import com.karl.constants.Globals;
+import com.karl.wrappers.Connectable;
+
+class HTTPRequestBuilder {
```
bob the builder over here very nice!

lpenzey commented on src/test/java/com/karl/ClientConnectionTest.java:
```diff
@@ -0,0 +1,45 @@
+    @Test
+    public void readUntilShouldGiveYouTheWholeStringIfLimiterIsNotFound() throws IOException {
+      String content =
+          "oh buddy its a string with a clrf oh no what's going to happen!!" + Globals.CRLF;
```
😆

lpenzey commented:
seriously good work, nice job Karl

ranizilpelwar commented on README.md:
````diff
@@ -1 +1,63 @@
+To run the HTTP server you could run
+
+```shell
+java -jar target/java-server-1.0-SNAPSHOT.jar 5000 http
````
Nice. I like that you give the specific command that you need to use to run the app or the tests, for example. 

A bonus would be to provide an example of the specific syntax you could type to make the Get call to echo back what you type in.

ranizilpelwar commented on README.md:
````diff
@@ -1 +1,63 @@
+mvn verify -DskipITs
+```
+
+So far I have not found a consistent way to only run the acceptance tests, as they need a server running in order for them to work you cannot just run the tests in isolation. This is less impactful then not being able to skip them in order for faster unit testing though.
````
I like that you are specific here on the current limitations of what you can run. This is insightful to the user reading your README.

ranizilpelwar commented on src/main/java/com/karl/ProtocolFactoryable.java:
```diff
@@ -0,0 +1,7 @@
+
+import com.karl.wrappers.Connectable;
+
+public interface ProtocolFactoryable {
```
Nice use of interface here!
Naming interfaces to start with the letter I will make a readability improvement when reading it from other classes. This is a useful convention.

ranizilpelwar commented on src/main/java/com/karl/ServerType.java:
```diff
@@ -0,0 +1,14 @@
+  public static ServerType fromString(String value) {
+    try {
+      return ServerType.valueOf(value);
+    } catch (IllegalArgumentException e) {
```
I like that you are thinking of the case where it may not match.
